### ISSUE TITLE
Unrepairable overflow! (See screenshot)

### ISSUE THREAD

**Issue by Moonbase59** [ErsatzTV/ErsatzTV#1465]
After everything ran for a while, the transition to next movie was due (playing a collection of HEVC movies for testing), and I got this error:

![Screenshot 2023-10-06 at 18-44-08 Jellyfin](https://github.com/ErsatzTV/ErsatzTV/assets/3706922/e843a2ce-cdf7-4f9f-a516-7c75c6e36bce)

Here’s the ETV media info (hopefully for the right movie this time…):

```
{
  "Id": 346,
  "Kind": "JellyfinMovie",
  "LibraryKind": "JellyfinLibrary",
  "ServerName": "media",
  "LibraryName": "Filme",
  "State": "RemoteOnly",
  "Duration": "01:22:55.6798976",
  "SampleAspectRatio": "1:1",
  "DisplayAspectRatio": "2.35:1",
  "RFrameRate": "24.00",
  "VideoScanKind": "Progressive",
  "Width": 1920,
  "Height": 816,
  "Streams": [
    {
      "Index": 0,
      "Kind": "Video",
      "Codec": "hevc",
      "Profile": "main 10",
      "Language": "eng",
      "Default": true,
      "Forced": true,
      "PixelFormat": "yuv420p10le",
      "ColorRange": "",
      "ColorSpace": "bt709",
      "ColorTransfer": "",
      "ColorPrimaries": ""
    },
    {
      "Index": 1,
      "Kind": "Audio",
      "Title": "English DD\u002B Stereo",
      "Codec": "eac3",
      "Profile": "",
      "Language": "eng",
      "Channels": 2,
      "Default": true
    },
    {
      "Index": 2,
      "Kind": "Audio",
      "Title": "Deutsch DD\u002B Stereo",
      "Codec": "eac3",
      "Profile": "",
      "Language": "ger",
      "Channels": 2
    },
    {
      "Index": 3,
      "Kind": "Subtitle",
      "Codec": "dvdsub",
      "Language": "eng"
    },
    {
      "Index": 4,
      "Kind": "Subtitle",
      "Codec": "dvdsub",
      "Language": "ger"
    },
    {
      "Index": 5,
      "Kind": "Subtitle",
      "Codec": "dvdsub",
      "Language": "ger",
      "Forced": true
    },
    {
      "Index": 6,
      "Kind": "Subtitle",
      "Codec": "subrip",
      "Language": "ger",
      "Forced": true
    }
  ]
}
```

(Note for myself: "Dr. Who und die Daleks")

The smart collection uses this search:
```
video_codec:hevc AND type:movie AND video_dynamic_range:sdr
```

**Comment by Moonbase59** [comment]
Just as I was typing this (after ~10-15 minutes of showing above error message on my TV), the movie suddenly _did_ start playing on my TV.

After another ~2 minutes, a new error is now shown:

![Screenshot 2023-10-06 at 19-03-12 Jellyfin](https://github.com/ErsatzTV/ErsatzTV/assets/3706922/13133785-1ef4-41fa-9df6-753c62cf29c0)

I `ssh`'ed into the media server and ETV is using a _lot_ of ffmpeg threads, and lots of CPU. Fortunately, we can differentiate between ETV and Jellyfin here, because Jellyfin use their own variant of ffmpeg, so `/usr/local/bin/ffmpeg` is _always_ ETV.

![matthias@media: ~_001](https://github.com/ErsatzTV/ErsatzTV/assets/3706922/e59becf0-1750-4301-bb7a-963dc4472db2)

**Comment by Moonbase59** [comment]
Restarted docker, supposedly other movie, same error, but this time showing an extreme load average and not much CPU%:

![matthias@media: ~-ersatztv_001](https://github.com/ErsatzTV/ErsatzTV/assets/3706922/184c82eb-7be4-4354-949b-609767f94ade)

**Comment by jasongdove** [comment]
Will dig into the error a bit more, but to provide context for the CPU use:

- HEVC decoding is CPU
- Primary content scaling is CPU (because decoding and later overlay are CPU, it doesn't make sense to go to hardware and back)
- Primary content padding is CPU
- Subtitle scaling is CPU
- Watermark scaling is CPU
- Subtitle overlay is CPU
- Watermark overlay is CPU

These are not inexpensive operations, so it's not surprising that CPU use is high.

It's worth considering that a higher resolution that more closely matches the source material may be more performant (i.e. not require scaling and padding) than a lower resolution.

**Comment by jasongdove** [comment]
Also, if you have a channel limited to h264 source content - I would try QSV instead of VAAPI acceleration in an FFmpeg Profile since the ffmpeg `overlay_qsv` filter doesn't appear to have the same bugs that prevent use of the `overlay_vaapi` filter, so more of the pipeline would probably use hardware.

I say limited to h264 because ETV's QSV capability detection (when to use hardware, when to use software) isn't as robust as with VAAPI.

**Comment by Moonbase59** [comment]
Thanks for caring! Yup, sure, doing everything in software is driving the CPUs to the limit. Expected, somehow, just looking for a way that might allow driving 2 or 3 clients (family) in case one of the few H.265 movies creeps up. They’re just too good to omiss, but I cant afford re-encoding 30-40 hours each (can’t afford buying newer hardware). Bad enough I have to re-encode 3D blurays for my TV…

Anyways, I tried QSV but that’s apparently not the way to go with my old Haswell i7-4790: Pegs 6 CPUs out @ ~90%, htop shows ffmpeg eating 500-550% CPU, and playout using browser (Jellyfin) stops after a few seconds, even with my 30s H.264 "filler" loop. So back to VAAPI/i965 for the time being.

Avoiding too much scaling might be a good idea, didn’t think about that—assumed 1280x720 target would use less resources. Then again, most of my sources are 1080p bluray rips, a few TV shows downloaded from official German media libraries ("Mediatheken") are 720p, and some exotic series (Get Smart)  ripped from old DVDs (480/576).

So I switched to a 1920x1080 h264 aac profile and am currently trying to find out what the best bitrate might be that allows 2 clients over the internet plus one here at home (TV), without too much dropouts and blocking artifacts. The suggested 2000 kBits/s produces too many blocking artifacts, mostly seen when a scene fades in from black. I have an upload of 35 MBits/s, so 4000 or 8000 might be okay maybe.

Using 8000 kBits/s on my Jellyfin AndroidTV client unfortunately produces a slight sub-1s frame loss every so many minutes, maybe something on the Amazon FireTV 4K stick can’t keep up, or it’s due to Jellyfin transcoding down for ATV, who knows.

I also see a short stuttering/dropout whenever a new client connects, can this have to do with the need of preparing a new stream?

This apparently leads to audio/video getting out of sync after a while.

Honestly, all these gazillion hardware limitations, drivers and options thing is slowly getting over my head—so many variables and such little real knowledge.

It’s amazing that guys like you actually cope with all that and manage to create things like ETV, KODI, Jellyfin… Chapeau.

Nevertheless, ETV and Jellyfin seem much too good and promising, so it’ll be worth pursuing this route. I think.
